### PR TITLE
Add RTCQuicStream interface. 

### PR DESCRIPTION
**Pull request by zhihuang0718**
Add RTCQuicStream interface. Add createStream() and onstream to RTCQu…icTransport.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -8640,16 +8642,17 @@
...
       <div>
         <pre class="idl">
-        [ Constructor (RTCceTransport transport, sequence&lt;RTCCertificate&gt; certificates), Exposed=Window]
+        [ Constructor (RTCIceTransport transport, sequence&lt;RTCCertificate&gt; certificates), Exposed=Window]
 interface RTCQuicTransport {
     readonly        attribute RTCIceTransport          transport;
     readonly        attribute RTCQuicTransportState    state;
     RTCQuicParameters     getLocalParameters ();
-    RTCQuicParameters?    getRemoteParameters ();
+    RTCQuicParameters     getRemoteParameters ();
```
Comment (pthatcherg) [inline index.html:8650]: Why did you change this?

File: index.html [hunk 2]
```diff
@@ -9142,6 +9161,252 @@
...
+      <h3>Overview</h3>
+      <p>An <code><a>RTCQuicStream</a></code> instance is associated to
+      an <code><a>RTCQuicTransport</a></code> instance.</p>
+    </section>
+    <section id="rtcquicstream-operation*">
+      <h3>Operation</h3>
+      An <code><a>RTCQuicStream</a></code> instance is created by the
+      <code><a>RTCQuicTransport</a></code> with a QUIC stream underneath.
+      When writing, the <code><a>RTCQuicStream</a></code>  will buffer
+      the data if the transport underneath is not writable. When receiving,
```
Comment (pthatcherg) [inline index.html:9178]: Should probably note that waitForWritable must be called before write() or write() will (likely) fail.

File: index.html [hunk 3]
```diff
@@ -9142,6 +9161,252 @@
...
+      the received data will be buffered and methods <code>read()</code>
+      and <code>readInto()</code> read the data from the buffer.</p>
```
Comment (pthatcherg) [inline index.html:9180]: Should probably not that waitForReadableOrFinished should be called before read() or it will likely fail or return only a very small amount.

File: index.html [hunk 4]
```diff
@@ -9142,6 +9161,252 @@
...
+    <section>
+      <h2>Attributes</h2>
+      <dl data-link-for="RTCQuicStream" data-dfn-for="RTCQuicStream" class=
+      "attributes">
+        <dt><dfn><code>onacked</code></dfn> of type <span class=
+        "idlAttrType"><a>EventHandler</a></span></dt>
+        <dd>
+          <p>This event handler, of event handler event type
+          <code><a>acked</a></code>, <em class="rfc2119" title="MUST">MUST</em>
+          be fired on reception of an ACK from the remote side.</p>
```
Comment (pthatcherg) [inline index.html:9210]: It's only fired once the stream has been finished and all the data in the entire stream has been acked.

File: index.html [hunk 5]
```diff
@@ -9142,6 +9161,252 @@
...
+    </section>
+    <section>
+      <h2>Methods</h2>
+      <dl data-link-for="RTCQuicStream" data-dfn-for="RTCQuicStream" class=
+      "methods">
+        <dt><code>write</code></dt>
+        <dd>
+          <p><dfn>write()</dfn> writes the data to local buffer and send the data when the
+          transport underneath is writeable. If the buffer is full, it will fail.</p>
+          <table class = "parameters">
```
Comment (pthatcherg) [inline index.html:9229]: Should say that one must call waitForWritable before calling write.   And we should be careful about using the word "writable" in the method waitForWritable, it means "there is room in the write buffer".

File: index.html [hunk 6]
```diff
@@ -9142,6 +9161,252 @@
...
+              </tr>
+            </tbody>
+          </table>
+          <div>
+            <em>Return type:</em> <code>void</code>
+          </div>
+        </dd>
+        <dt><code>finish</code></dt>
+        <dd>
+          <p><dfn>finish()</dfn> will maybe wirte one last value before finishing and then a "fin" bit.
```
Comment (pthatcherg) [inline index.html:9255]: wirte => write

Comment (pthatcherg) [inline index.html:9255]: Not "maybe write" but "write one last value (if given) and then write the QUIC fin bit"

File: index.html [hunk 7]
```diff
@@ -9142,6 +9161,252 @@
...
+            </tbody>
+          </table>
+          <div>
+            <em>Return type:</em> <code>void</code>
+          </div>
+        </dd>
+        <dt><code>waitForWritable</code></dt>
+        <dd>
+          <p><dfn>waitForWritable()</dfn> will be resolved when the amount, up to maxBufferedAmount,
+          can be written or buffered. If the maxBufferedAmount is not provided, it will be equal
```
Comment (pthatcherg) [inline index.html:9283]: It will be resolved when "amount" is available in the buffer, and the max buffered amount is maxBufferedAmount.   In other words, it waits until the buffer is drained if it's already too full.  That's not quite clear in this text as-is.

File: index.html [hunk 8]
```diff
@@ -9142,6 +9161,252 @@
...
+          </table>
+          <div>
+            <em>Return type:</em> <code>void</code>
+          </div>
+        </dd>
+        <dt><code>waitForReadableOrFinished</code></dt>
+        <dd>
+          <p><dfn>waitForReadableOrFinished()</dfn> will be resolved when the amount, up to
+          maxBufferedAmount, can be read. If the maxBufferedAmount is not provided, it will
+           be equal to the amount.</p>
```
Comment (pthatcherg) [inline index.html:9360]: It will be resolved when "amount" is available in the buffer, and the max buffered amount is maxBufferedAmount. In other words, it will keep filling up to maxBufferedAmount, even if it's more than "amount" (it can keep filling even after the promise resolves).

File: index.html [hunk 9]
```diff
@@ -10323,6 +10588,33 @@ <h3 id="sec.identity-verify-assertion">Verifying Identity Assertions</h3>
...
+          <th>Event name</th>
+          <th>Interface</th>
+          <th>Fired when...</th>
+        </tr>
+      </tbody>
+      <tbody>
+        <tr>
+          <td><code>ack</code></td>
+          <td><code><a>Event</a></code></td>
+          <td>Received an ACK from the remote side</td>
```
Comment (pthatcherg) [inline index.html:10611]: Again, it's when *all* data has been acked (and after fin has been sent)

**Review by pthatcherg — COMMENTED** [review]
Thanks for writing this!

We should make a PR agains the main repository.

**Comment by taylor-b** [comment]
Sorry for taking too long to look at this; it looks like Bernard already added a pretty similar API: https://github.com/w3c/ortc/pull/785

The main difference is that it uses a state enum, and doesn't have the "waitFor" or "finish" methods.
